common/modules/nft_deposit/src/nft_deposit.rs

```diff
@@ -0,0 +1,133 @@
+pub trait NftDepositModule: token_send::TokenSendModule + token_supply::TokenSupplyModule {
+    fn deposit_tokens(
+        &self,
+        payment_token_id: TokenIdentifier,
+        payment_token_nonce: Nonce,
+        payment_amount: Self::BigUint,
+    ) -> SCResult<()> {
+        require!(payment_amount != 0, "Cannot deposit 0 tokens");
+        require!(payment_token_nonce != 0, "Cannot deposit fungible tokens");
+        require!(
+            self.nft_deposit_accepted_token_ids()
+                .contains(&payment_token_id),
+            "Not an accepted token id"
+        );
+
+        let payment = GenericEsdtAmountPair {
+            token_id: payment_token_id,
+            token_nonce: payment_token_nonce,
+            amount: payment_amount,
+        };
+
+        let mut entry_updated = false;
+        let caller = self.blockchain().get_caller();
+        let mut deposit = self.nft_deposit(&caller).get();
+
+        for entry in deposit.iter_mut() {
+            if self.equal_token_type(entry, &payment) {
+                entry.amount += &payment.amount;
+                entry_updated = true;
+                break;
+            }
+        }
+
+        if !entry_updated {
+            require!(
+                deposit.len() < self.nft_deposit_max_len().get(),
+                "Deposit is full"
+            );
+            deposit.push(payment);
+        }
+
+        self.nft_deposit(&caller).set(&deposit);
+        Ok(())
+    }
+
+    #[endpoint(withdrawAllTokensFromDeposit)]
+    fn withdraw_all_tokens_from_deposit(
+        &self,
+        #[var_args] opt_accept_funds_func: OptionalArg<BoxedBytes>,
+    ) -> SCResult<()> {
+        let caller = self.blockchain().get_caller();
+        let deposit = self.nft_deposit(&caller).get();
+
+        self.nft_deposit(&caller).clear();
+        deposit.iter().for_each(|entry| {
+            self.send_nft_tokens(
+                &entry.token_id,
+                entry.token_nonce,
+                &entry.amount,
+                &caller,
+                &opt_accept_funds_func,
+            )
+        });
+
+        Ok(())
+    }
+
+    #[endpoint(withdrawTokenFromDeposit)]
+    fn withdraw_token_from_deposit(
+        &self,
+        index: usize,
+        #[var_args] opt_accept_funds_func: OptionalArg<BoxedBytes>,
+    ) -> SCResult<()> {
+        let caller = self.blockchain().get_caller();
+        let mut deposit = self.nft_deposit(&caller).get();
+        require!(index > 0 && index < deposit.len(), "Index out of range");
+
+        let entry = deposit.remove(index);
+        self.nft_deposit(&caller).set(&deposit);
+
+        self.send_nft_tokens(
+            &entry.token_id,
+            entry.token_nonce,
+            &entry.amount,
+            &caller,
+            &opt_accept_funds_func,
+        );
+
+        Ok(())
+    }
+
+    fn burn_deposit_tokens(
+        &self,
+        caller: &Address,
+        deposit: &[GenericEsdtAmountPair<Self::BigUint>],
```

Comment:
The argument is just a copy of the storage. Here the storage itself must be cleared

Reply:
I think you did something like
```

Reply:
Because reading a vector from storage might be gas expensive. I observed that every time when burning the tokens the deposit is already loaded, so passed it as parameter here to save some gas. It is a tradeoff between clean code and gas efficiency. I'd say the gas is more important in this case.

Reply:
ok